File: algorithm/data_structure/list/linked_list.py
# -*- coding: utf-8 -*-


class Node:
    def __init__(self, data):
        self.next = None
        self.data = data


class LinkedList:
    def __init__(self):
        self.head = None

    def length(self):
        length = 0
        current = self.head

        while current is not None:
            current = current.next
            length += 1

        return length

    def insert(self, index, data):
        new_node = Node(data)
        tmp = self.head

        if index == 0:
            self.head = new_node
            new_node.next = tmp
        else:
            i = 0
            while (tmp is not None) and (i < index):
                prev = tmp
                tmp = tmp.next
                i += 1

            prev.next = new_node
            new_node.next = tmp

    def delete(self, index):
        deleted = None

        if index == 0:
            deleted = self.head.data
            self.head = self.head.next
        else:
            tmp = self.head
            i = 0
            while (tmp is not None) and (i < index):
                prev = tmp
                tmp = tmp.next
                i += 1

            deleted = tmp.data
            prev.next = tmp.next

        return deleted

    def delete_all(self):
        self.head = None

    def print_all(self):
        current = self.head

        print('[ ', end='')
        while current is not None:
            print('{} '.format(current.data), end='')
            current = current.next
        print(']')


def main():
    linked_list = LinkedList()
    linked_list.insert(0, 1)
    linked_list.print_all()

    linked_list.insert(1, 3)
    linked_list.print_all()

    linked_list.insert(1, 9)
    linked_list.print_all()

    length = linked_list.length()
    print('Length: {}\n'.format(length))

    linked_list.delete(1)
    linked_list.print_all()

    linked_list.insert(2, 5)
    linked_list.print_all()

    linked_list.insert(100, 8)
    linked_list.print_all()

    linked_list.delete_all()
    length = linked_list.length()
    print('Length: {}\n'.format(length))
    linked_list.print_all()


if __name__ == '__main__':
    main()

File: algorithm/data_structure/list/linked_list.cpp
#include <iostream>

using namespace std;

struct Node
{
    int value;
    Node* next;
};

Node* create()
{
    Node* head = new Node;
    head->next = NULL;
    return head;
}

void destroy(Node* list)
{
    Node* current = list;
    Node* tmp;

    while (current != NULL) {
        tmp = current->next;
        delete current;
        current = tmp;
    }
}

void insert(Node** head, int index, int value)
{
    Node* node = new Node;
    node->value = value;
    node->next = NULL;

    if (index == 0) {
        Node* tmp = *head;
        *head = node;
        node->next = tmp;
    } else {
        Node* previous = *head;
        Node* current = previous->next;
        int i = 1;
        while ((i < index) && (current != NULL)) {
            previous = current;
            current = current->next;
            ++i;
        }
        previous->next = node;
        node->next = current;
    }
}

void remove(Node** head, int index)
{
    if (index == 0) {
        Node* tmp = *head;
        *head = tmp->next;
        delete tmp;
    } else {
        Node* previous = *head;
        Node* current = previous->next;
        int i = 1;
        while ((i < index) && (current != NULL)) {
            previous = current;
            current = current->next;
            ++i;
        }
        previous->next = current->next;
        delete current;
    }
}

void print_list(Node* head)
{
    Node* current = head;
    while (current != NULL) {
        cout << current->value << ' ';
        current = current->next;
    }
    cout << '\n';
}

int main()
{
    Node* head = create();
    print_list(head);

    insert(&head, 0, 10);
    print_list(head);

    insert(&head, 2, 5);
    print_list(head);

    insert(&head, 1, 7);
    print_list(head);

    remove(&head, 0);
    print_list(head);

    remove(&head, 1);
    print_list(head);

    remove(&head, 1);
    print_list(head);

    destroy(head);
}

